topic_dynamics/parsing.py

```diff
@@ -0,0 +1,420 @@
+    return files
+
+
+def read_file(file: str) -> bytes:
+    """
+    Read the contents of the file.
+    :param file: address of the file.
+    :return: bytes with the contents of the file.
+    """
+    with open(file) as fin:
+        content = bytes(fin.read(), "utf-8")
+    return content
+
+
+def get_positional_bytes(node: tree_sitter.Node) -> Tuple[int, int]:
+    """
+    Extract start and end byte.
+    :param node: node on the AST.
+    :return: (start byte, end byte).
+    """
+    start = node.start_byte
+    end = node.end_byte
+    return start, end
+
+
+def get_identifiers(file: str, lang: str) -> List[Tuple[str, int]]:
+    """
+    Gather a sorted list of identifiers in the file and their count.
+    :param file: address of the file.
+    :param lang: the language of file.
+    :return: a list of tuples, identifier and count.
+    """
+    content = read_file(file)
+    tree = get_parser(lang).parse(content)
+    root = tree.root_node
+    identifiers = []
+
+    def traverse_tree(node: tree_sitter.Node) -> None:
+        """
+        Run down the AST (DFS) from a given node and gather identifiers from its children.
+        :param node: starting node.
+        :return: None.
+        """
+        for child in node.children:
+            if child.type in NODE_TYPES[lang]:
+                start, end = get_positional_bytes(child)
+                identifier = content[start:end].decode("utf-8").lower()
+                if "\n" not in identifier:  # Will break output files. Can add other bad characters later
+                    identifiers.append(identifier)
+            if len(child.children) != 0:
+                traverse_tree(child)
+
+    traverse_tree(root)
+    sorted_identifiers = sorted(Counter(identifiers).items(), key=itemgetter(1), reverse=True)
+
+    return sorted_identifiers
+
+
+def transform_identifiers(identifiers: List[Tuple[str, int]]) -> List[str]:
+    """
+    Transform the original list of identifiers into the writable form.
+    :param identifiers: list of tuples, identifier and count.
+    :return: a list of identifiers in the writable form, "identifier:count".
+    """
+    formatted_identifiers = []
+    for identifier in identifiers:
+        if identifier[0].rstrip() != "":  # Checking for occurring empty tokens.
+            formatted_identifiers.append(identifier[0].rstrip() + ":" + str(identifier[1]).rstrip())
+    return formatted_identifiers
+
+
+def slice_and_parse(repository: str, output_dir: str, dates: List[datetime.datetime], lang: str, name: str) -> None:
+    """
+    Split the repository, parse the full files, write the data into a file.
+    Can be called for parsing full files and for parsing diffs only.
+    When run several times, overwrites the data.
+    :param repository: path to the repository to process.
+    :param output_dir: path to the output directory.
+    :param dates: a list of dates used for slicing.
+    :param lang: language of parsing.
```

Comment:
```suggestion
    :param lang: programming language to use.
```
Should it be multiple languages?

Reply:
Will be removed with enry.